AutoBackup Functionality with Auto Shutdown

KrX3D opened the pull request:
This adds AutoBackup functionality in the LuckyBackup Docker container by adding:

# New environment variables:
AUTOBACKUP: Enables automatic backup when the container starts.
AUTOSHUTDOWN: Powers off the container after the backup completes.
PROFILENAME: Specifies the backup profile to be used.

# Improved setup instructions:
Users are advised to set AUTOBACKUP=false on first run to configure their tasks before enabling automatic backups.
Detailed steps on modifying the container settings to allow shutdown after backup (--restart=no).

# Simplified automation:
Users can now trigger a backup simply by starting the container, which will run the backup and then shut itself down when finished.

## This is particularly useful for automation setups like Unraid User Scripts, where a script can:
Power on a Synology NAS via WOL.
Start the LuckyBackup container to run the backup.
Automatically shut down the NAS after the backup is complete and the container stops.
This enhancement makes automated, on-demand backups more streamlined and power-efficient. 🚀

ich777 commented:
@KrX3D first of all thank you for the PR!

Please remove your GitHub Workflows since I don't use them and they are not part of my workflow.

Do you use the container on Unraid or on another system? I assume you are using the container on Synology.
I just reviewed the PR and it seems rather complicated for a user to set up since he has to remove `Extra Parameters` and alongside with adding other variables.

What you describe is also easily achievable by just writing a script and run that on startup on the host, something like:
```
sleep 2m
docker start <CONTAINERNAME>
sleep 30
docker exec -i <CONTAINERNAME> /usr/bin/luckybackup --no-questions /luckybackup/.luckyBackup/profiles/<PROFILENAME>.profile
```

I'm also using a while loop in the script to check for a control file that is written lastly by the backup profile so that the script knows when the backup is finished and shuts down the backup machine.

So to speak what the script does is:

1. Wait 2 minutes
2. Start the container
3. Wait 30 seconds to give the container time to fully start
4. Execute the backup job in the container

I'll update the script above when I get home. I assume you can run a script on a Synology NAS on boot too?

KrX3D commented:
Hi, 


> @KrX3D first of all thank you for the PR!

Nothing to thank me, i need to thank you for creating it
 
> Please remove your GitHub Workflows since I don't use them and they are not part of my workflow.
> 
i removed the workflow, i needed it since i didnt know another way how i can use and test it.



> Do you use the container on Unraid or on another system? I assume you are using the container on Synology. I just reviewed the PR and it seems rather complicated for a user to set up since he has to remove `Extra Parameters` and alongside with adding other variables.

No im using unraid for this docker, the backup is made from unraid to synology. The extra Parameter can be changed in the docker contrainer settings in unraid:

![Unbenannt](https://github.com/user-attachments/assets/ec692929-75f2-423d-b9da-dfa45aeca326)

Just toggle Advanced View to on and than you can see it. it can probably be set up in the dockerfile? but i didnt want to change it, in case you just use it your way

with other variables you mean those ?:

AUTOBACKUP,AUTOSHUTDOWN and PROFILENAME

since i added them in the dockerfile they where present for he also in the contrainer settings and ther you just need to set it to true if you want to use it,:

![Unbenannt2](https://github.com/user-attachments/assets/6186be06-5923-4b27-8c42-753c69707f25)

PS: the Proflle name "default" is also the default name in luckybackup, so it the user did not change it he doesnt need to change it there too.
The script also checks it and if the name is incorrect the automatic backup is skipped
   -> echo "---Error: Profile ${PROFILENAME} not found! Skipping backup and shutting down in 10 seconds for debugging---"


> What you describe is also easily achievable by just writing a script and run that on startup on the host, something like:
> 
> ```
> sleep 2m
> docker start <CONTAINERNAME>
> sleep 30
> docker exec -i <CONTAINERNAME> /usr/bin/luckybackup --no-questions /luckybackup/.luckyBackup/profiles/<PROFILENAME>.profile
> ```

really? didnt know about that, but i still think thats easier the the way i did it,

> I'm also using a while loop in the script to check for a control file that is written lastly by the backup profile so that the script knows when the backup is finished and shuts down the backup machine.
> 
> So to speak what the script does is:
> 
> 1. Wait 2 minutes
> 2. Start the container
> 3. Wait 30 seconds to give the container time to fully start
> 4. Execute the backup job in the container
> 
> I'll update the script above when I get home. I assume you can run a script on a Synology NAS on boot too?

Like i said i also use unraid so i didnt test it, since i cant even use docker on my synology nas, to old

ich777 commented:
> The extra Parameter can be changed in the docker contrainer settings in unraid

Sure I know that. ;)

> Just toggle Advanced View to on and than you can see it. it can probably be set up in the dockerfile? 

Not, that won't work in the `Dockerfile`.

> since i added them in the dockerfile they where present for he also in the contrainer settings and ther you just need to set it to true if you want to use it

Just to clarify, just because you've added them to the `Dockerfile` they won't be automatically be added to the template in Unraid, users would have to add them manually, however what is true of course if you set `AUTOBACKUP` and `AUTOSHUTDOWN` by default to `false` users won't have to worry about that since it will be disabled and not visible on existing installations.

> really? didnt know about that, but i still think thats easier the the way i did it,

Yes, it's possible, I'll send you the link when I have a bit more time and I'm at home.

However none of the methods above are "easy" to set up for a average user because my script method involves the Plugin `User Scripts` but after that it will be just copy paste.

Your method of course involves the risk if a user configures something wrong that it will simply shutdown the server.

I'm also a friend of making things easier for users but over simplification can be also involve causing more troubles than it solves, at least that's what my experience is.

ich777 commented:
@KrX3D this is basically my script which runs as a User Script on startup from the Server, you can also see there that I use Pushbits (a Notoficiation service for Matrix-Synapse) to push messages and notify me about the status, this could be of course be changed to support Discord or whatever Push service you are using, even Email should be possible:
```
#!/bin/bash
# Wait 30 seconds to give all services time to start
sleep 30

# Set Pushbits variables
URL=<ADDRESS>
TOKEN=<TOKEN>

# Set push_message function
push_message() {
curl \
    -X POST "$URL/message?token=$TOKEN" \
    -F "title=$1" \
    -F "message=$2" \
    >/dev/null 2>&1
}

# Push message that backup will start in 2 minutes and wait 2 minutes
push_message "<SERVERNAME> Backup" "Backup from <SERVERNAME> to <BACKUPSERVERNAME> will start in 2 minutes!"
sleep 120

# Check if directory for control file exists
if [ ! -d /tmp/luckybackup ]; then
  mkdir /tmp/luckybackup
fi

# Send push message
push_message "<SERVERNAME> Backup starting" "Backup from <SERVERNAME> to <BACKUPSERVERNAME> starting!"

# Execute default luckybackup task
docker exec luckyBackup bash -c "/usr/bin/luckybackup -c --no-questions /root/.luckyBackup/profiles/default.profile > /root/.luckyBackup/logs/default-LastCronLog.log 2>&1"

# Check for control file every minute
while true; do
  if [ -f "/tmp/luckybackup/control" ]; then
    break
  else
    sleep 60
  fi
done

# Remove control file
rm -rf mkdir /tmp/luckybackup/control

# Push message that backup is finished
push_message "<SERVERNAME> Backup finished" "Backup from <SERVERNAME> to <BACKUPSERVERNAME> finished. <BACKUPSERVERNAME> Server will shutdown in one minute!"

# Wait 60 seconds and shutdown Backup Server
sleep 60
powerdown
```

As said above, I mounted /tmp/luckybackup to the container and as the last sync job I simply have a directory with a `control` file which syncs to that directory so that the script knows when everything is done and shutdown the server.

This basically means you could turn on the server, might be even remotely, wait for all the messages and it will automatically shut down the server again.
Since this is running on the host rather than inside the container you can do all kinds of fancy stuff like the push messages and shutdown from the server.

I also don‘t see what you mean with save power by shuting down the container since if no one is connected to the web interface and no backup is running it won‘t use any power or CPU cycles.

KrX3D commented:
> > since i added them in the dockerfile they where present for he also in the contrainer settings and ther you just need to set it to true if you want to use it
> 
> Just to clarify, just because you've added them to the `Dockerfile` they won't be automatically be added to the template in Unraid, users would have to add them manually, however what is true of course if you set `AUTOBACKUP` and `AUTOSHUTDOWN` by default to `false` users won't have to worry about that since it will be disabled and not visible on existing installations.

you sure, i didnt add anything and those 3 extra parameters just poped up in my templete

and yes setting them to default false would be good.

> Your method of course involves the risk if a user configures something wrong that it will simply shutdown the server.

which server do you mean? just by miss configuring the container should just shutdown the container.


> @KrX3D this is basically my script which runs as a User Script on startup from the Server,

so if i understand it right you power on your server (real machine) and that runs the usersctipt and backs it up?

than arent we talking about 2 different things?
my unraid server (real machine) is running 24/7 and each 2 days a user sctipt powers on via WOL my synology nas which is just used for backup
after it is powered on the user sctipt powers on the luckybackup container which after it finishes all tasks shuts itself down
and lastly after that happened the usersctipt poowers off my synology nas (via ssh)

so my synology nas just powers on for around 30 minutes each 2 days to get all backups from my unraid server

my userscript looks like this:

```
#!/bin/bash

# Synology NAS details
NAS_IP="X.X.X.X"
NAS_MAC="XX:XX:XX:XX:XX:XX"

# LuckyBackup container name
DOCKER_CONTAINER="luckyBackup"

echo "Sende Wake-on-LAN Signal an die Synology NAS..."
etherwake -i br0 $NAS_MAC

#echo "Warte, bis die Synology NAS per Ping antwortet..."
#while ! ping -c 1 -W 1 "$NAS_IP" &> /dev/null; do
    #echo "      Warte auf Ping-Antwort..."
    #sleep 5
#done

#echo "Ping ist OK, warte auf vollständigen Systemstart (SSH-Port 22)..."
echo "Warte auf vollständigen Systemstart (SSH-Port 22)..."
while ! (echo > /dev/tcp/"$NAS_IP"/22) &> /dev/null; do
    echo "      Warten auf SSH-Port..."
    sleep 5
done
echo "SSH-Port ist erreichbar, warte 30 Sekunden bis Synology vollständig gestartet ist."

sleep 30
echo "Synology NAS ist online."

# Starte den LuckyBackup Docker-Container
echo "Starte LuckyBackup Container..."
docker start $DOCKER_CONTAINER

# Warte, bis der LuckyBackup-Container gestoppt wird
echo "Warte, bis LuckyBackup die Sicherung abgeschlossen hat..."
while docker ps --format '{{.Names}}' | grep -q "^$DOCKER_CONTAINER$"; do
    sleep 60
done
echo "LuckyBackup-Container wurde gestoppt."

# Synology NAS herunterfahren
echo "Fahre die Synology NAS herunter..."
ssh -i /mnt/user/appdata/luckybackup/.ssh/id_rsa SYNOLOGY_USER@$NAS_IP "sudo /sbin/poweroff"

echo "Warte, bis die Synology NAS per Ping antwortet..."
while ! ping -c 1 -W 1 "$NAS_IP" &> /dev/null; do
    echo "      Warte auf Ping-Antwort..."
    sleep 5
done

echo "Backup-Prozess abgeschlossen."
```

nevermind that i used ssh instead of ping to wait till the sysnology powers up, this  just worked better for me

> As said above, I mounted /tmp/luckybackup to the container and as the last sync job I simply have a directory with a `control` file which syncs to that directory so that the script knows when everything is done and shutdown the server.
> 
> This basically means you could turn on the server, might be even remotely, wait for all the messages and it will automatically shut down the server again. Since this is running on the host rather than inside the container you can do all kinds of fancy stuff like the push messages and shutdown from the server.

so again, you talk about your backup server (unraid) which powers on, does the backup and powers off right?


> I also don‘t see what you mean with save power by shuting down the container since if no one is connected to the web interface and no backup is running it won‘t use any power or CPU cycles.

you are talking about this:
This enhancement makes automated, on-demand backups more streamlined and power-efficient. 🚀

? i meant that since my synology nas is only running for the time the backup is done, before it was running much longer. so my bad about not clarifying it

ich777 commented:
> you sure, i didnt add anything and those 3 extra parameters just poped up in my templete

Trust me, that was not the case, definitely not because Unraid doesn't add the Variables from the Dockerfile automatically <- this is done for security reasons.



> so if i understand it right you power on your server (real machine) and that runs the usersctipt and backs it up?
> 
> than arent we talking about 2 different things?

Exactly.



> nevermind that i used ssh instead of ping to wait till the sysnology powers up, this just worked better for me

This will also work.



> so again, you talk about your backup server (unraid) which powers on, does the backup and powers off right?

Exactly but the script could also be modified to just stop the container or send a shutdown through SSH.


However I have to say after looking again at this all, this is nothing I want to support in my container.
You can of course maintain your own fork of the container and even possibly put it into the CA App if you want to keep that functionality since it's all open source. :)

KrX3D commented:
@ich777 

i tested it your way and this is the unraid userscript i use now (just the part with thr docker, before and after i power on/off my synology nas like posted in the comment above):

```#!/bin/bash

# LuckyBackup container name
DOCKER_CONTAINER="luckyBackup"

# Starte den LuckyBackup Docker-Container
echo "Starte LuckyBackup Container..."
docker start $DOCKER_CONTAINER

sleep 30

# Generate timestamp for log file
TIMESTAMP=$(date +"%Y-%m-%d_%H-%M-%S")
LOG_FILE="/luckybackup/.luckyBackup/logs/default-LastCronLog-$TIMESTAMP.log"

echo "Backup auf Synology wird gestartet..."
docker exec $DOCKER_CONTAINER bash -c "/usr/bin/luckybackup -c --no-questions /luckybackup/.luckyBackup/profiles/default.profile > $LOG_FILE 2>&1"

sleep 10
docker stop --time=15 $DOCKER_CONTAINER || docker kill $DOCKER_CONTAINER
echo "LuckyBackup-Container wurde gestoppt."
```

This creates the default-LastCronLog-$TIMESTAMP.log, the TIMESTAMP part can be removed but i like to keep each log

but i needed to to set " Run as root user" to true in the docker settings or else the other logs for each task werent created (some permission problem)


it would be nice if you could add this or your way into the readme, if i had known before i would not have created a fork.

ich777 commented:
> but i needed to to set " Run as root user" to true in the docker settings or else the other logs for each task werent created (some permission problem)

Change it to that if you don't run it as root:
```
docker exec --user luckybackup $DOCKER_CONTAINER bash -c "/usr/bin/luckybackup -c --no-questions /luckybackup/.luckyBackup/profiles/default.profile > $LOG_FILE 2>&1"
```
